File: src/main/java/org/jetbrains/kotlinx/lincheck/strategy/randomswitch/RandomSwitchStrategy.java
package org.jetbrains.kotlinx.lincheck.strategy.randomswitch;

/*
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2015 - 2018 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import org.jetbrains.kotlinx.lincheck.execution.*;
import org.jetbrains.kotlinx.lincheck.runner.*;
import org.jetbrains.kotlinx.lincheck.strategy.*;
import org.jetbrains.kotlinx.lincheck.verifier.*;

import java.lang.reflect.*;
import java.util.*;

import static org.jetbrains.kotlinx.lincheck.strategy.LincheckFailureKt.toLincheckFailure;


/**
 * This managed strategy switches the current thread to a random one with the specified probability.
 * In addition it tries to avoid both communication and resource deadlocks and to check for livelocks.
 * <p>
 * TODO: not developed yet, dummy implementation only
 */
public class RandomSwitchStrategy extends ManagedStrategy {
    private final int invocations;
    private final Verifier verifier;

    public RandomSwitchStrategy(RandomSwitchCTestConfiguration testCfg, Class<?> testClass,
                                ExecutionScenario scenario, List<Method> validationFunctions,
                                Verifier verifier) {
        super(testClass, scenario, validationFunctions);
        this.invocations = testCfg.invocationsPerIteration;
        this.verifier = verifier;
    }

    @Override
    protected LincheckFailure runImpl() {
        for (int i = 1; i < invocations; i++) {
            InvocationResult ir = runInvocation();
            if (ir instanceof CompletedInvocationResult) {
                ExecutionResult results = ((CompletedInvocationResult) ir).getResults();
                if (!verifier.verifyResults(getScenario(), results))
                    return new IncorrectResultsFailure(getScenario(), results);
            } else {
                return toLincheckFailure(ir, getScenario());
            }
        }
        return null;
    }

    @Override
    public void onStart(int iThread) {
        super.onStart(iThread);
    }

    @Override
    public void beforeSharedVariableRead(int iThread, int codeLocation) {
        Thread.yield();
    }

    @Override
    public void beforeSharedVariableWrite(int iThread, int codeLocation) {
        Thread.yield();
    }
}

File: src/main/java/org/jetbrains/kotlinx/lincheck/strategy/randomswitch/RandomSwitchStrategy.kt
/*-
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2019 JetBrains s.r.o.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */
package org.jetbrains.kotlinx.lincheck.strategy.randomswitch

import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario
import org.jetbrains.kotlinx.lincheck.runner.CompletedInvocationResult
import org.jetbrains.kotlinx.lincheck.strategy.IncorrectResultsFailure
import org.jetbrains.kotlinx.lincheck.strategy.LincheckFailure
import org.jetbrains.kotlinx.lincheck.strategy.ManagedStrategyBase
import org.jetbrains.kotlinx.lincheck.strategy.modelchecking.ModelCheckingCTestConfiguration
import org.jetbrains.kotlinx.lincheck.strategy.toLincheckFailure
import org.jetbrains.kotlinx.lincheck.verifier.Verifier
import java.lang.reflect.Method

private const val startSwitchProbability = 0.05
private const val endSwitchProbability = 0.99

/**
 * RandomSwitchStrategy just switches at every code location to a random thread.
 * For internal experiments only!
 */
internal class RandomSwitchStrategy(
        testCfg: RandomSwitchCTestConfiguration,
        testClass: Class<*>,
        scenario: ExecutionScenario,
        validationFunctions: List<Method>,
        verifier: Verifier
) : ManagedStrategyBase(
        testClass, scenario, verifier, validationFunctions, ModelCheckingCTestConfiguration.DEFAULT_HANGING_DETECTION_THRESHOLD,
        false, ArrayList(ModelCheckingCTestConfiguration.DEFAULT_GUARANTEES)
) {
    // the number of invocations that the managed strategy may use to search for an incorrect execution
    private val maxInvocations = testCfg.invocationsPerIteration
    private var switchProbability = startSwitchProbability

    @Throws(Exception::class)
    override fun runImpl(): LincheckFailure? {
        repeat(maxInvocations){
            // switch probability changes linearly from startSwitchProbability to endSwitchProbability
            switchProbability = startSwitchProbability + it * (endSwitchProbability - startSwitchProbability) / maxInvocations
            when (val ir = runInvocation()) {
                is CompletedInvocationResult -> {
                    if (!verifier.verifyResults(scenario, ir.results))
                    return IncorrectResultsFailure(scenario, ir.results)
                }
                else -> return ir.toLincheckFailure(scenario)
            }
        }
        return null
    }

    override fun shouldSwitch(threadId: Int): Boolean {
        // TODO: can reduce the number of random calls using geometric distribution
        return random.nextDouble() < switchProbability
    }

    override fun chooseThread(switchableThreads: Int): Int = random.nextInt(switchableThreads)

    override fun initializeInvocation() {
        super.initializeInvocation()
        // start from random thread
        currentThread = random.nextInt(nThreads)
    }
}